### DISCUSSION TITLE
Is it possible to send data from a child machine to parent?

### DISCUSSION THREAD

**Discussion by panesofglass** [statelyai/xstate#1445] [General] [answered]
I am attempting to share data between child machines through the parent's context. However, I'm having a hard time determining how to pass data from a child machine back to the parent machine without using a `final` state, which I don't want to use. Is there a `data` option that can be used with `sendParent`, or do I need to use an `EventObject` to pass data?

Original:
I'm currently invoking child machines from a parent machine to orchestrate several, related workflows. I would like these to remain available, as it is possible to jump back and forth, and I want to retain the last state of each. However, I also need to share state from the children with one another. I know I can use a final state and pass data through `onDone`, but I'd like to not do that. I could potentially `spawn` the machine each time and initialize it with the parent context. Is there another way?

**Comment by Andarist** [comment, marked as answer]
I'm not sure if I have understood correctly because the content of the issue doesn't match the title as far as I can tell. 

But if you just want to send something from a child to a parent then `sendParent` should be what you are looking for. If you want to send an event to a sibling then you can also do this, but you have to share target's `sessionId` first - which can be done through events. That `sessionId` is always set as `_event.origin`.

**Reply by panesofglass** [reply]
Thank you, @Andarist. I updated my question. The first option of using `sendParent` is what I've tried, but for some reason, I get a errors when trying to send an event object. Perhaps I just need to add additional typings to the event object I try to send to make TypeScript happy.

The second option is interesting, but I don't think I've seen either `sessionId` or `_event.origin` in any docs or logged objects. Do you have an example to which you could link?

**Reply by Andarist** [reply]
> Thank you, @Andarist. I updated my question. The first option of using sendParent is what I've tried, but for some reason, I get a errors when trying to send an event object. Perhaps I just need to add additional typings to the event object I try to send to make TypeScript happy.

Yeah, u might need to provide an extra type param or smth. If you prepare a runnable repro case then I could help you with that.

> The second option is interesting, but I don't think I've seen either sessionId or _event.origin in any docs or logged objects. Do you have an example to which you could link?

You can get it from the third argument given to your actions/conds/others: https://github.com/davidkpiano/xstate/blob/73f5802d86d5689d367281c37c4f0aaee7003450/packages/core/test/event.test.ts#L26-L27
